src/lib/storage/create_iterable_from_segment.hpp

```diff
@@ -25,85 +38,36 @@ class ReferenceSegmentIterable;
-    return ValueSegmentIterable<T>{segment};
-  }
-}
+auto create_iterable_from_segment(const ValueSegment<T>& segment);
 
 template <typename T, bool EraseSegmentType = HYRISE_DEBUG>
-auto create_iterable_from_segment(const DictionarySegment<T>& segment) {
-#ifdef HYRISE_ERASE_DICTIONARY
-  PerformanceWarning("DictionarySegmentIterable erased by compile-time setting");
-  return AnySegmentIterable<T>(DictionarySegmentIterable<T, pmr_vector<T>>(segment));
-#else
-  if constexpr (EraseSegmentType) {
-    return create_any_segment_iterable<T>(segment);
-  } else {
-    return DictionarySegmentIterable<T, pmr_vector<T>>{segment};
-  }
-#endif
-}
+auto create_iterable_from_segment(const DictionarySegment<T>& segment);
 
 template <typename T, bool EraseSegmentType = HYRISE_DEBUG>
-auto create_iterable_from_segment(const RunLengthSegment<T>& segment) {
-#ifdef HYRISE_ERASE_RUNLENGTH
-  PerformanceWarning("RunLengthSegmentIterable erased by compile-time setting");
-  return AnySegmentIterable<T>(RunLengthSegmentIterable<T>(segment));
-#else
-  if constexpr (EraseSegmentType) {
-    return create_any_segment_iterable<T>(segment);
-  } else {
-    return RunLengthSegmentIterable<T>{segment};
-  }
-#endif
-}
+auto create_iterable_from_segment(const RunLengthSegment<T>& segment);
 
 template <typename T, bool EraseSegmentType = HYRISE_DEBUG>
-auto create_iterable_from_segment(const FixedStringDictionarySegment<T>& segment) {
-#ifdef HYRISE_ERASE_FIXEDSTRINGDICTIONARY
-  PerformanceWarning("FixedStringDictionarySegmentIterable erased by compile-time setting");
-  return AnySegmentIterable<T>(DictionarySegmentIterable<T, FixedStringVector>(segment));
-#else
-  if constexpr (EraseSegmentType) {
-    return create_any_segment_iterable<T>(segment);
-  } else {
-    return DictionarySegmentIterable<T, FixedStringVector>{segment};
-  }
-#endif
-}
+auto create_iterable_from_segment(const FixedStringDictionarySegment<T>& segment);
 
-template <typename T, bool EraseSegmentType = HYRISE_DEBUG>
-auto create_iterable_from_segment(const FrameOfReferenceSegment<T>& segment) {
-#ifdef HYRISE_ERASE_FRAMEOFREFERENCE
-  PerformanceWarning("FrameOfReferenceSegmentIterable erased by compile-time setting");
-  return AnySegmentIterable<T>(FrameOfReferenceSegmentIterable<T>(segment));
-#else
-  if constexpr (EraseSegmentType) {
-    return create_any_segment_iterable<T>(segment);
-  } else {
-    return FrameOfReferenceSegmentIterable<T>{segment};
-  }
-#endif
+template <typename T, typename Enabled, bool EraseSegmentType = HYRISE_DEBUG>
+auto create_iterable_from_segment(const FrameOfReferenceSegment<T, Enabled>& segment);
+
+// Fix template deduction so that we can call `create_iterable_from_segment<T, false>` on FrameOfReferenceSegments
+template <typename T, bool EraseSegmentType, typename Enabled>
+auto create_iterable_from_segment(const FrameOfReferenceSegment<T, Enabled>& segment) {
+  return create_iterable_from_segment<T, Enabled, EraseSegmentType>(segment);
```

Comment:
Couldn't this line be moved into the .ipp file?

Reply:
I guess so, but since the ipp file is not its own compilation unit, it doesn't change anything for the compiler. Thus, for the sake of readability, I preferred not to have another forward declaration.